Enhancement/tool class restructuring

vlagrassa opened the pull request:
Well, I don't think I hit the "remove more lines than you add" goal...

(More detailed breakdown coming)

vlagrassa committed:
Indentation

vlagrassa committed:
Organize Entity subclass imports

vlagrassa committed:
Reorder Species & Wormbase imports to remove local import in IndelFinder

vlagrassa committed:
Refactor tool report & task name(s)

Report classes are now `<Tool>Report`, and Task classes are `<Tool>Task`.

- `IndelFinder` -> `IndelFinderReport`
- `NemascanMapping` -> `NemascanReport`
- `NemaScanTask` -> `NemascanTask`
- `MappingSubmissionManager` -> `NemascanSubmissionManager`

vlagrassa committed:
Create JobPipeline and Runner template classes

Pretty major restructuring of job pipeline. Jobs now represented by a `JobPipeline` subclass, which is a manager for the following:
- A `Report` class, which controls storage
- A `Task` class, which controls scheduling
- A `Runner` class, which controls running

Code previously in `SubmissionManager` has been moved to `JobPipeline` and its subclasses, most noticeably the bulk of the submission logic into `JobPipeline.create()`.

Code previously in `TaskHandler` has almost entirely been moved to `Runner` and its subclasses, which are now considered part of the data model (instead of a service).

Job creation (making a Report / Entity and uploading input file(s)) has been separated from job scheduling (creating and submitting a Task).  As part of this, I've made a new `TaskStatus` value `"CREATED"`, to represent jobs that have been created in cloud storage but not yet scheduled.

vlagrassa committed:
Rename tools/submit.py to validate.py

vlagrassa committed:
Bugfix: JobPipeline import

vlagrassa committed:
Combine 'create_job' and 'run_job' using 'start_job' from pipeline utils

vlagrassa committed:
Abstract Runner class away from specific Report

Instances of the `Runner` class or its subclasses are considered unique up to `data_id`.
For the tool jobs, this is the data hash; for database operations, this is the operation name (e.g. `TEST_ECHO`).

To run a job, a specific report must be provided.  This will fill out the template provided by the runner class with variables specific to that report.

vlagrassa committed:
Pass valid file extensions to parse function directly

vlagrassa committed:
Create new db op using JobPipeline create & schedule

vlagrassa committed:
Move db op preflight check to pipeline parsing step, make DbOp enum

vlagrassa committed:
Move queue name lookup to JobPipeline module

Lookup is now based directly on the Task subclasses and their `queue` name values

vlagrassa committed:
Produce execution ID & pipeline op record inside Runner run() function

The `Runner.run` function now:
- handles the creation of a `PipelineOperation` entity recording the new execution
- returns a unique ID for the job execution that was started

vlagrassa committed:
Pull container from report when starting job instead of saving at Runner init

vlagrassa committed:
Fix: define _Record_Class instead of just annotating type

vlagrassa committed:
Rename container_name to job_name and use explicit DNS_LABEL conversion function

vlagrassa committed:
Rework Runner class initialization

- `kind` is now an abstract property and must be defined by any instantiable subclass
- At the basic level, a `Runner` can be created using exactly one of:
  - Data ID (saves value as-is)
  - Report (must have correct `kind`, and valid `data_id` in the specified field

vlagrassa committed:
Make pipeline_subclasses frozen set

vlagrassa committed:
Handle status check in Runner class

Runner now includes abstract method `check_status`, which can be used to check the status of a specific execution of the job.

The status check also handles updating the `PipelineOperation` record, so the task `/status` route doesn't have to explicitly update this anymore.
More concretely, the code for updating the record has been moved into the `GCPRunner` and `GCPCloudRunRunner` classes.

Doing things this way has highlighted the possibility of job name collisions. Since job names are composed of `{kind}-{data_id}`,
it's technically possible for two different jobs to have the same name if the `kind` of one is a prefix of the other.
This isn't possible with any of our currently existing `kind`s, and in fact is mitigated if the `data_id` cannot contain hyphens.
Some of the code in this commit is to try and reasonably future-proof this solution for just this problem.

vlagrassa committed:
Validate report kind & data ID before running

vlagrassa committed:
Update a few comments

vlagrassa committed:
Refactor TaskStatus to JobStatus

The status is more accurately a property of the pipeline as a whole, rather than the task specifically.

This has the nice effect of moving the definition out to the `services` level, so we no longer have to locally import the status in Entity methods to avoid circular imports.

Also added CSS color support for the new `CREATED` value, between report creation & task submission.

vlagrassa committed:
Fix: Store operation name in job record

This step is handled by the `JobPipeline` class, since it mediates the interactions between the component classes (in this case, the Report and Runner classes).

Promotes `get_full_execution_name` to a method on the base `Runner` class, to make it available in `JobPipeline`.

vlagrassa committed:
Keep PipelineOperation record private to Runner class

The `check_status` function now explicitly produces a `JobStatus` value.

Refactored `update_all_linked_status_records` to take the `Runner`, the execution ID, and the new status. This keeps the `PipelineOperation` record completely contained inside the `Runner` class.

vlagrassa committed:
Define safe update status using JobPipeline object

vlagrassa committed:
Fix: Use report data ID when warning about re-run

vlagrassa committed:
Include comment for potential header validator functions

This is pretty low priority, but if we were to accept files without fixed headers but with rules about those headers' formats, we could do it like this.

vlagrassa committed:
Delete obsolete create_pipeline_operation_record function

This is now handled implicitly in the `Runner` subclass(es).

vlagrassa committed:
Clean up error usage in run function

vlagrassa committed:
Use DbOp enum to validate database operation values

For now, store as a plain string after validation because the Enum class doesn't save to GCP out of the box. (Proximate error is that it's not JSON serializable.)

TODO: Support enum entity properties?

vlagrassa committed:
Move LocalRunner to its own file

vlagrassa committed:
Separate base Runner class and GCP subclasses into separate files

vlagrassa committed:
Fix: GCPCloudRunRunner import

VSCode didn't save the file, so it didn't make it into the previous commit... :/

vlagrassa committed:
Merge branch 'development' into enhancement/tool-class-restructuring

vlagrassa committed:
Fix: request to /create must include operation ID

vlagrassa commented on src/modules/api/pipeline-task/pipelines/task_handler.py:
Most of this code has been moved to the new `Runner` abstract base class and its subclasses.

In particular, the code for creating/running a CloudRun job has been moved to `GCPCloudRunRunner`, and the (deprecated) code for running a Lifesciences pipeline has been moved to `GCPLifesciencesRunner`.

vlagrassa commented on src/modules/api/pipeline-task/pipelines/utils.py:
```diff
@@ -36,131 +26,89 @@ def load_json_from_request(request):
     raise APIBadRequestError('Failed to parse request body as valid JSON') from ex
 
 
-def log_ssl_backoff(details):
```
Moved to `GCPCloudRunRunner`

vlagrassa commented on src/modules/api/pipeline-task/pipelines/utils.py:
```diff
@@ -36,131 +26,89 @@ def load_json_from_request(request):
-
-
 
 def get_task_handler(queue_name, *args, **kwargs):
```
Now returns an object of the appropriate `JobPipeline` subclass, instead of a `TaskHandler`

vlagrassa commented on src/modules/api/pipeline-task/pipelines/utils.py:
```diff
@@ -36,131 +26,89 @@ def load_json_from_request(request):
-@backoff.on_exception(
-    backoff.constant, SSLEOFError, max_tries=3, interval=20, jitter=None, on_backoff=log_ssl_backoff, on_giveup=log_ssl_giveup,
-)
-def start_job(payload, task_route, run_if_exists=False):
```
Converted into the implementation for the `run` abstract method in `GCPCloudRunRunner`

vlagrassa commented on src/modules/api/pipeline-task/pipelines/utils.py:
```diff
@@ -36,131 +26,89 @@ def load_json_from_request(request):
 
-
-def update_status_safe(queue_name, op_id, call_id='', status=None, **kwargs):
+def update_status_safe(job: JobPipeline, status: JobStatus, call_id: str):
```
By using the `JobPipeline` object directly, we can avoid defining a bunch of job-specific update functions in the `service` subdirectory.

vlagrassa commented on src/modules/api/pipeline-task/pipelines/utils.py:
```diff
@@ -36,131 +26,89 @@ def load_json_from_request(request):
-    backoff.constant, SSLEOFError, max_tries=3, interval=20, jitter=None, on_backoff=log_ssl_backoff, on_giveup=log_ssl_giveup,
-)
-def start_job(payload, task_route, run_if_exists=False):
+def get_runner_from_operation_name(operation_name: str) -> Tuple[GCPCloudRunRunner, Optional[str]]:
```
New function

vlagrassa commented on src/modules/api/pipeline-task/routes/task.py:
```diff
@@ -37,43 +37,29 @@ def start_task(task_route):
+  handler = get_task_handler(task_route, **payload)
 
-
-  # Create a Pipeline Operation record for the task
```
The diff is a bit misleading here.  Creating & updating the `PipelineOperation` execution record is now handled implicitly by the `GCPRunner` class and its subclasses.  The new code uses the `Runner` subclass's `run` method to start the job, instead of the old (removed) function `start_job`.

vlagrassa commented on src/modules/api/pipeline-task/routes/task.py:
```diff
@@ -94,11 +80,10 @@ def update_task():
+  logger.debug(f"[{ call_id }] Retrieving the operation...")
   try:
-    op = update_pipeline_operation_record(operation_name)
+    runner, exec_id = get_runner_from_operation_name(operation_name)
```
Since the `PipelineOperation` record is now handled implicitly by the `GCPRunner` subclass, we instead parse the operation name into a `Runner` and an execution ID.

vlagrassa commented on src/modules/api/pipeline-task/routes/task.py:
```diff
@@ -107,13 +92,22 @@ def update_task():
-    logger.debug(f"[{ call_id }] Updating all linked status records for operation {op}: {dict(op)}")
-    update_all_linked_status_records(op['operation_kind'], op['operation'])
+    logger.debug(f"[{ call_id }] Updating all linked status records to status { status }...")
+    update_all_linked_status_records(runner, exec_id, status)
```
The `update_all_linked_status_records` function has been rewritten to use the `Runner` directly, instead of the `PipelineOperation`.  In a future PR, this is a good candidate for a class method on `JobPipeline` -- looking up all `Report`s that match a given `Runner` and updating them.

vlagrassa commented on src/modules/api/pipeline-task/routes/task.py:
```diff
@@ -128,4 +122,4 @@ def update_task():
   # If the job is still running, don't acknowledge -- tells Pub/Sub to try the request again
   # Return as bool value to be handled by pubsub_endpoint decorator
-  return op['done'] or op['error']
+  return status in JobStatus.FINISHED
```
Since the `PipelineOperation` is now managed entirely by `GCPRunner`, we instead compare the resulting status to the set of "finished" status values (which in my opinion is a better solution anyways).

vlagrassa commented on src/modules/site-v2/base/utils/tools.py:
```diff
@@ -58,63 +63,77 @@ def lookup_report(kind, reportId, user=None, validate_user=True):
+def try_submit(kind, user, data, no_cache):
 
-  # Try submitting the job
   try:
```
As part of the new pipeline architecture, `create`ing and `schedule`-ing a job are now separate processes.  Looking forward, this will allow us to define the Phenotype report as createable but not schedulable, since the results are computed in-place.

vlagrassa commented on src/modules/site-v2/base/utils/tools.py:
```diff
@@ -58,63 +63,77 @@ def lookup_report(kind, reportId, user=None, validate_user=True):
     }, 200
 
   # Duplicate job submission from another user
   except CachedDataError as ex:
```
I'll admit this is slightly hack-y, but this error can only be thrown if the `job` object was instantiated correctly, so we have access to it in this branch (but not in others).  There might be a better way to rewrite this whole process without using errors, but I've left it as-is for now since it does work.

vlagrassa commented on src/modules/site-v2/base/views/admin/etl.py:
```diff
@@ -94,21 +97,18 @@ def create_op():
+
+  # Try parsing the form into a job handler & submitting the job
+  try:
+    job = DatabaseOperationPipeline.create(user, form.data)
```
I've reanalyzed the `db_op_preflight_check` code as part of the parsing step for a `DatabaseOperationPipeline` job, so the code has been moved to that class (i.e. no need to explicitly call a check here).

vlagrassa commented on src/pkg/caendr/caendr/models/datastore/__init__.py:
I've re-ordered and organized the imports here to make the dependencies more explicit (which allowed me to remove some local imports).

vlagrassa commented on src/pkg/caendr/caendr/models/job_pipeline/database_operation_pipeline.py:
Most of this (`REQUIRED_FILES` and `parse` in particular) was moved over from `services/cloud/database_operation.py`.  The old preflight check function is now captured in the `DatabaseOperationPipeline.parse` method.

vlagrassa commented on src/pkg/caendr/caendr/models/job_pipeline/heritability_pipeline.py:
Most code moved from old `HeritabilitySubmissionManager` class.

vlagrassa commented on src/pkg/caendr/caendr/models/job_pipeline/indel_primer_pipeline.py:
Most code moved over from old `IndelPrimerSubmissionManager` class

vlagrassa commented on src/pkg/caendr/caendr/models/job_pipeline/nemascan_pipeline.py:
Most code moved over from old `NemascanSubmissionManager` class.

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+    pass
+
+
+  def _create_execution_record(self, execution_id, response):
```
Moved from pipeline utils

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+  # Container Command #
+
+  @abstractmethod
+  def construct_command(self, report: DataJobEntity) -> list:
```
Moved from `TaskHandler`

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+      "GOOGLE_ZONE":                  GOOGLE_CLOUD_ZONE,
+    }
+  
+  def get_data_job_vars(self, report: DataJobEntity):
```
Moved from `TaskHandler`

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+
+  # Container Environment #
+
+  def get_gcp_vars(self):
```
Moved from `TaskHandler`

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+  _Record_Class = PipelineOperation
+
+  # Machine Parameters #
+  # Can be overwritten in subclasses as needed
```
Machine Parameters and CloudRun Parameters moved from `TaskHandler`

vlagrassa committed:
Move GCP service specific parameters to appropriate subclasses

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+    op_id = self.get_full_execution_name(execution_id)
+    status = get_job_execution_status(op_id)
+
+    # Try looking up and updating the pipeline operation record
```
Moved from pipeline utils (or maybe cloud utils)

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+
+  #
+  # Backoff Logging Functions
+  #
```
Moved from `pipelines/utils.py`

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+  @backoff.on_exception(
+      backoff.constant, SSLEOFError, max_tries=_MAX_TRIES_TOTAL, interval=20, jitter=None, on_backoff=_log_ssl_backoff.__func__, on_giveup=_log_ssl_giveup.__func__,
+  )
+  def run(self, report, run_if_exists=False):
```
Moved from `pipelines/utils.py`

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+  @backoff.on_exception(
+      backoff.expo, HttpError, giveup=lambda ex: ex.status_code != 400, max_tries=_MAX_TRIES_RUN, on_backoff=_log_backoff.__func__, on_giveup=_log_giveup.__func__, jitter=None
+  )
+  def _run(self):
```
Moved from `TaskHandler`

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+  # Creating & Running CloudRun Job
+  #
+
+  def _create(self, report: DataJobEntity):
```
Moved from `TaskHandler`

vlagrassa commented on src/pkg/caendr/caendr/models/run/gcp.py:
```diff
@@ -0,0 +1,525 @@
+  # def check_status(self):
+  #   return get_pipeline_status(self.operation_name)
+
+  def run(self, report, run_if_exists=False):
```
Moved from `TaskHandler`

vlagrassa commented on src/pkg/caendr/caendr/models/run/local.py:
```diff
@@ -0,0 +1,8 @@
+from .base import Runner
+
+
+class LocalRunner(Runner):
```
We can fill this out & use it for the phenotype tool

vlagrassa commented on src/pkg/caendr/caendr/models/run/tools_gcp_cloudrun.py:
Almost entirely moved from `TaskHandler` subclasses

vlagrassa commented on src/pkg/caendr/caendr/models/run/tools_gcp_cloudrun.py:
```diff
@@ -0,0 +1,164 @@
+  # Names #
+
+  @classmethod
+  def split_job_name(cls, job_name: str) -> (str, str):
```
Probably the only new function in this file.

vlagrassa commented on src/pkg/caendr/caendr/models/status.py:
Moved out from `task.py` -- was originally called `TaskStatus`, but I think `JobStatus` is more appropriate.

vlagrassa commented on src/pkg/caendr/caendr/models/task.py:
```diff
@@ -16,31 +16,6 @@
 
 ## Task Superclass Definition ##
 
-# Define the Status values
```
Moved to `JobStatus`

vlagrassa commented on src/pkg/caendr/caendr/services/cloud/lifesciences.py:
```diff
@@ -47,36 +46,6 @@ def start_pipeline(SERVICE, task_id, pipeline_request):
     raise PipelineRunError(err)
 
 
-def create_pipeline_operation_record(task, response):
```
Moved to `GCPCloudRunRunner.run` method. The `PipelineOperation` execution record is now handled implicitly by that class.

vlagrassa commented on src/pkg/caendr/caendr/services/cloud/utils.py:
```diff
@@ -24,6 +24,34 @@
 
 
 
+def make_dns_name_safe(s: str) -> str:
```
Abstracted out from `TaskHandler`. It technically doesn't do everything specified by the format -- might have to update this.

vlagrassa commented on src/pkg/caendr/caendr/services/cloud/utils.py:
```diff
@@ -50,83 +78,38 @@ def get_operation_status(operation_name):
 
 
 
-def update_pipeline_operation_record(operation_name):
```
Moved to `GCPRunner` class (or maybe its subclass).

vlagrassa commented on src/pkg/caendr/caendr/services/cloud/utils.py:
```diff
@@ -50,83 +78,38 @@ def get_operation_status(operation_name):
+  op_name = runner.get_full_execution_name(exec_id)
+  logger.debug(f'[UPDATE {op_id}] Updating all linked status records with kind = {runner.kind}, operation name = {op_name} to status "{status}"')
 
-  # Get the operation status, plus the parsed service & id values
```
The conversions here are handled by the `Runner` subclass.

vlagrassa commented on src/pkg/caendr/caendr/services/database_operation.py:
```diff
@@ -20,51 +16,6 @@
 
 
 
-DB_OPS = {
```
Names and titles moved to `models/datastore/DatabaseOperation.py` in the `DbOp` enum, required files moved to the `DatabaseOperationPipeline` class as part of the preflight check in `parse`

vlagrassa commented on src/pkg/caendr/caendr/services/database_operation.py:
```diff
@@ -128,87 +79,4 @@ def get_etl_op(op_id, keys_only=False, order=None, placeholder=True):
-  return [(key, val['title']) for key, val in DB_OPS.items()]
-
-
-def db_op_preflight_check(op, species_list):
```
Moved to `DatabaseOperationPipeline` as part of the `parse` method

vlagrassa commented on src/pkg/caendr/caendr/services/database_operation.py:
```diff
@@ -128,87 +79,4 @@ def get_etl_op(op_id, keys_only=False, order=None, placeholder=True):
-  return missing_files
-
-
-def create_new_db_op(op, user, args=None, note=None):
```
Now handled by the standard `JobPipeline.create` method

vlagrassa commented on src/pkg/caendr/caendr/services/heritability_report.py:
```diff
@@ -50,23 +49,6 @@ def get_heritability_reports(username=None, filter_errs=False):
 
 
 
-def update_heritability_report_status(id: str, status: str=None, operation_name: str=None):
```
Handled by the `JobPipeline` subclass

vlagrassa commented on src/pkg/caendr/caendr/services/indel_primer.py:
```diff
@@ -176,23 +175,6 @@ def query_indels_and_mark_overlaps(species, strain_1, strain_2, chromosome, star
 
 
 
-def update_indel_primer_status(id: str, status: str=None, operation_name: str=None):
```
Handled by the `JobPipeline` subclass

vlagrassa commented on src/pkg/caendr/caendr/services/nemascan_mapping.py:
```diff
@@ -52,29 +49,8 @@ def get_mappings(username=None, filter_errs=False):
-
-
-
-def update_nemascan_mapping_status(id: str, status: str=None, operation_name: str=None):
```
Handled by the `JobPipeline` subclass

vlagrassa commented on src/pkg/caendr/caendr/services/tools/submit.py:
Submission and specific validation code moved to `JobPipeline` and its subclasses.

Validator functions (`validate_file`, `validate_strain`, etc) moved to `validate.py` file.  (I thought GitHub tracked this as a file rename...)

vlagrassa commented on src/pkg/caendr/caendr/services/validate.py:
Moved from `submit.py`.  I thought GitHub tracked this as a file rename...

r-vieira commented on src/pkg/caendr/caendr/models/run/__init__.py:
```diff
@@ -0,0 +1,7 @@
+# Parent classes
+from .base   import Runner
+from .gcp    import GCPRunner, GCPCloudRunRunner, GCPLifesciencesRunner
```
Unless we still need LifeSciences logic, we should plan on retiring all references to it.

vlagrassa replied:
Agreed. I was hesitant to delete it entirely, so I've tried to modularize it as much as possible and remove remaining imports.

I removed one of the imports in [f7000d5](https://github.com/AndersenLab/CAENDR/pull/404/commits/f7000d5ddc10d0a41cd49606718b93bb61510c40).

There are a few remaining references I could find:
- The dedicated `lifesciences.py` files, in `services/cloud/` and in `models/`
- The `PipelineRunError` class in `error.py` (which holds almost every error in the app, I think)
- The `cloud/utils.py` function `get_operation_status`, which routes the a status check to the correct function (lifesciences or cloudrun, in our case)
- The `GCPLifesciencesRunner` class.  This is marked as deprecated, and intentionally doesn't implement all of its parent's abstract methods, so in practice it can't be instantiated.

I'm happy to move `GCPLifesciencesRunner` to its own file and write "deprecated" everywhere, but I'd like to do that in a new branch if possible -- I have some incoming changes to the `Runner`s that I'd like to PR first, and I'd like to avoid merge conflicts.

r-vieira commented on src/pkg/caendr/caendr/services/nemascan_mapping.py:
```diff
@@ -2,19 +2,16 @@
 
 
-def is_data_cached(ns: NemascanMapping):
+def is_data_cached(ns: NemascanReport):
```
not sure if I misread it earlier, but we should standardize capitalization of `Nemascan` vs `NemaScan`

vlagrassa replied:
Agreed -- I've changed all instances in the code to `Nemascan`, with lowercase `'s'`.  There are still a few text instances that use `NemaScan`, (e.g. the text on the tool page, env vars & GCP paths), but I'm not sure we can refactor those the same way.

r-vieira reviewed (changes requested):
Looks terrific! Please review minor inline comments.

vlagrassa committed:
Move `get_operation_id_from_name` out of lifesciences file

This function is more general than just lifesciences, so I've moved it to the `cloud/utils.py` file.

vlagrassa commented:
Working on merging w `development` -- there were a few updates to `submit.py`, which is deleted in this branch, so it will take a sec to distribute the new code to the appropriate locations.

(If only GitHub had tracked it as a file rename instead of a raw add/delete... ;-;)

vlagrassa committed:
Merge branch 'development' into enhancement/tool-class-restructuring

Involved manually moving changes from `services/tools/submit.py` to `services/validate.py` and `models/job_pipeline/nemascan_pipeline.py`

vlagrassa committed:
Fix circular import (cloud utils & lifesciences)

For now, just using a local import... we're deprecating lifesciences anyways, so I'm not sure it's worth it to spend much more time trying to make this function work properly

vlagrassa committed:
Fix indentation
